### ISSUE TITLE
Make diff flags configurable

### ISSUE THREAD

**Issue by Drup** [ocaml/dune#697]
The builtin diff with automatic promotion is really cool, but it doesn't support custom diff flags. Some flags are very useful (ignoring whitespaces and so on), so it would be nice to be able to change them easily.

**Comment by rgrinberg** [comment]
Note that custom diffing is supported with `-diff-cmd` but what we'd like is a way to customize diffing per action (or even using profiles).

**Comment by mlasson** [comment]
Maybe there's a simple trick I am not aware of, but having an option to ignore the "\r"  before "\n" is a portability issue for executing tests on windows platforms.

**Comment by diml** [comment]
BTW, Dune first compares the two files by opening them in binary mode and comparing their contents. It only calls the diff command when they don't match, so the behavior is well defined and platform independent.

**Comment by mlasson** [comment]
The "Custom Test" idiom presented in the manual forces the tests to call "Pervasives.set_binary_mode_out stdout true" in order to work on linux and windows. 

I think there should be a simple way to make simple non-regression tests like that; but I have no idea what solution should be promoted. 
   0. use set_binary_mode_out, 
   1. we could add an alternative "diff" action that compare files upto \r\n ~ \n, 
   2. we could add an alternative redirection "with-stdout-to" action that normalize \r, 
   3. once pipe (Issue #428) is implemented, we could pipe the result into a "normalizer".

**Comment by rgrinberg** [comment]
@diml all the windows specific issues have been fixed, right? Can we summarize what is left doing for custom promotion.

**Comment by diml** [comment]
For Windows, yes. I had something in mind for custom promotion, but I can't remember it now

**Comment by vogler** [comment]
I run into this issue here: https://github.com/ocaml/opam-repository/pull/15897
I would like to ignore whitespace for `diff` but arguments for `--diff-command` do not seem to be supported:
~~~
$ dune runtest --diff-command='diff -w'
Error: command reported no differences: cd _build/default && diff -w
test.expected.ml test.actual.ml
~~~

**Comment by diml** [comment]
Indeed. Dune compares the files itself and only calls the diff command if it finds out that they are different. What you can do is process the output to make it more canonical.

**Comment by vogler** [comment]
Thanks, that's what I did in the referenced commit above.
Just to clearify the part "only calls the diff command if it finds out that they are different":
The files were different, I also made some non-whitespace change to test.
`dune runtest --diff-command=diff` worked, but `dune runtest --diff-command='diff -w'` results in the above error. So, I assume the problem is passing arguments for diff-command.

**Comment by diml** [comment]
Basically, if dune thinks the file are different it expect the diff command to print a diff and return a non-zero exit code. If it returns a zero exit code, dune thinks the diff command considers the file as equal and it report an error because the diff command doesn't agree with dune.

What is surprising is that `diff -w` didn't report an error for non-whitespace change.

**Comment by diml** [comment]
I mean didn't exit with a non-zero exit code.

**Comment by vogler** [comment]
If called directly, it does:
~~~
$ diff -w test.actual.ml test.expected.ml; echo $?
14c14
<   | (A x,_) when x <> 1 -> 1
---
>   | (A x, _) when x <> 0 -> 1
1
~~~
Passing arguments is actually not a problem:
~~~
$ dune runtest --diff-command='diff -w'
          sh (internal) (exit 1)
(cd _build/default && /bin/sh -c 'diff -w test.expected.ml test.actual.norm.ml')
15c15
<   | (_,A x) when x <> 1 -> 1
---
>   | (_, A x) when x <> 0 -> 1
~~~
I must have mixed up the tests.
As you say, the problem only occurs when there is only whitespace difference because then `diff -w`'s exit code is 0 while the internal diff sees a difference.

**Comment by yawaramin** [comment]
@mlasson 

> Maybe there's a simple trick I am not aware of, but having an option to ignore the "\r" before "\n" is a portability issue for executing tests on windows platforms.

See https://github.com/ocaml/dune/issues/8358#issuecomment-2585856282

**Comment by Leonidas-from-XIV** [comment]
We'll close this for now since most issues are fixed and in general I would suggest we'd view diffing as a process in Dune and not an invocation of a particular implementation overall (which also potentially allows us to use an integrated ocaml diff command which would work identical between GNU/BSD/Windows). If there's any specific things to be added, please reopen or create a new issue.
